solver/bboltcachestorage/storage.go

```diff
@@ -23,9 +26,31 @@ type Store struct {
 }
 
 func NewStore(dbPath string) (*Store, error) {
-	db, err := bolt.Open(dbPath, 0600, nil)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to open database file %s", dbPath)
+	fallback := true
+	var db *bolt.DB
+	for {
```

Comment:
This for loop isn't the most straightforward. My understanding of this code is that it will attempt to open it once, do a rename if it fails, then create a new database and there's extra logic with the fallback variable to prevent this being an infinite loop.

I think it might be more straightforward to refactor this out of the `NewStore` method and then implement the fallback using a more simple early return with no error or implement the fallback as part of the recover.

So something like:

```go
func safeOpenDBWithFallback(dbPath string) (*bbolt.DB, error) {
    db, err := safeOpenDB(dbPath)
    if err == nil {
        return db, err
    }

    // implement logging and fallback logic
}
```

Or implement a fallback function like this:

```go
func fallbackOpenDB(dbPath string, db **bbolt.DB, err *error) {
    r := recover()
    if r == nil {
        return
    }

    // log error, perform rename, then open db again
    *db, *err = bbolt.Open(dbPath, ...)
}

func safeOpenDB(dbPath string) (db *bbolt.DB, err error) {
    defer fallbackOpenDB(dbPath, &db, &err)
    return bolt.Open(dbPath, 0600, nil)
}
```

One advantage of the fallback function is that it _only_ runs when a panic occurs while the other code (and this code) runs on any error. So if the error is a user-error such as attempting to open an already locked database, you might just attempt to rename it and replace it.

Reply:
The purpose of the loop was to avoid calling `bolt.Open` or `safeOpenDB` multiple times. I don't think this is possible with the suggested patterns. If that is not the case then we could just remove the `for` loop and replace it with one extra `if safeOpenDB(); err != nil` call in the current implementation

Reply:
My opinion is that it would be better to call `bolt.Open` multiple times because it's more explicit. One of these calls is the primary call, then the backup, then opening the fallback. It encapsulates the logic in a more straightforward way and is more obvious what is happening. In the current iteration of the code, it does the same thing, but in a way that's more obfuscated.

And containing each one of those steps into their own named function would also make it more obvious what each step is intended to do and reduce the overall complexity of the code. Since this is a recovery section of the code, I think reducing complexity is more valuable than reducing lines of code.

Reply:
> reducing complexity is more valuable than reducing lines of code.

It's more about making sure that the fallback and original call are always using the same config and avoid possible problems where they could diverge in the future where only one side is changed and duplicate code path does something different.